oda_api/token.py

```diff
@@ -66,6 +65,34 @@ def decode_oauth2_token(token: str):
     # usually comes in cookies['_oauth2_proxy']
     return json.loads(base64.b64decode(token.split(".")[0]+"=").decode())
 
+
+def rewrite_token(new_token,
+                  token_discovery_methods=(
```

Comment:
I am not sure discovery is applicable when the token is being written.
When it's being found, it find the first usable one.

When writing, there are other options.
Among them, I think it's good to explicitly indicate where to store the token, not just bool.

Also, when token is being overwritten, report what is the difference between old and new token. Maybe even save the old token.

Reply:
> I am not sure discovery is applicable when the token is being written. When it's being found, it find the first usable one.
> 
> When writing, there are other options. Among them, I think it's good to explicitly indicate where to store the token, not just bool.
>

You mean having the user explicitly specify where the new token will be written?

Reply:
Yes. `write_token=False`, `write_token=token.HOME_FILE`, etc.

Also, when overwriting, to check if the new one is actually better. I.e. has at least the same roles and at least the same remaining lifetime.

Reply:
The refresh_token method allows now to specify where to store the new token.

Also, the old token is stored and afterwards the previous "location" used is emptied, in order to guarantee that the new, refreshed, token, get found properly.

Reply:
> I am not sure discovery is applicable when the token is being written. When it's being found, it find the first usable one.
> 

The token to refresh can now be passed as argument, if that's not the case, it will be searched and discovered. Also to me it makes more sense.

> 
> Also, when token is being overwritten, report what is the difference between old and new token. Maybe even save the old token.

A dedicated comparison function now actually checks the fields of the payloads of the old token and the refreshed one, differences are reported and, if necessary, the force the writing.